llm/llama.cpp/generate_cuda_linux.go

```diff
@@ -0,0 +1,103 @@
+//go:build exclude
```

Comment:
Building these separate runners is showing the limits of how we use `go generate`. There ends up being a lot of layers of indirection, `go generate -> go logic -> build and run shell commands from go`. Let's take this opportunity to simplify and move this logic to a shell script, I whipped this up based on your logic in here:
```
#!/bin/sh

# Set ROCM path from environment variable or default to /opt/rocm
ROCM_PATH="${ROCM_PATH:-/opt/rocm}"

# Check for CUDA and generate for GGUF and GGML
if nvcc &>/dev/null; then
    echo "Building with CUDA"

    cmake -S gguf -B gguf/build/cuda -DLLAMA_K_QUANTS=on -DLLAMA_ACCELERATE=on -DLLAMA_CUBLAS=on
    cmake --build gguf/build/cuda --target server --config Release

    cmake -S ggml -B ggml/build/cuda -DLLAMA_K_QUANTS=on -DLLAMA_ACCELERATE=on -DLLAMA_CUBLAS=on
    cmake --build ggml/build/cuda --target server --config Release

# Check for ROCM and generate for GGUF
elif [ -d "$ROCM_PATH" ]; then
    echo "Building with ROCM"

    # GGML does not support ROCM, only generate GGUF
    cmake -S gguf -B gguf/build/cuda -DLLAMA_K_QUANTS=on -DLLAMA_ACCELERATE=on -DLLAMA_HIPBLAS=on -DCMAKE_C_COMPILER=$ROCM_PATH/llvm/bin/clang -DCMAKE_CXX_COMPILER=$ROCM_PATH/llvm/bin/clang++
    cmake --build gguf/build/cuda --target server --config Release

# Check for OpenCL and generate for GGUF and GGML
elif clinfo &>/dev/null; then
    echo "Building with OpenCL"

    cmake -S gguf -B gguf/build/cuda -DLLAMA_K_QUANTS=on -DLLAMA_ACCELERATE=on -DLLAMA_CLBLAST=on
    cmake --build gguf/build/cuda --target server --config Release

    cmake -S ggml -B ggml/build/cuda -DLLAMA_K_QUANTS=on -DLLAMA_ACCELERATE=on -DLLAMA_CLBLAST=on
    cmake --build ggml/build/cuda --target server --config Release

else
    echo "No GPU found, skipping GPU build"
fi
```

The in the `generate_linux.go` we can just call this shell script for now and keep the cuda building logic in the shell script:
```
diff --git a/llm/llama.cpp/generate_linux.go b/llm/llama.cpp/generate_linux.go
index 4cae384..d0d7be0 100644
--- a/llm/llama.cpp/generate_linux.go
+++ b/llm/llama.cpp/generate_linux.go
@@ -16,7 +16,4 @@ package llm
 //go:generate cmake -S gguf -B gguf/build/cpu -DLLAMA_K_QUANTS=on
 //go:generate cmake --build gguf/build/cpu --target server --config Release
 
-//go:generate cmake -S ggml -B ggml/build/cuda -DLLAMA_CUBLAS=on -DLLAMA_ACCELERATE=on -DLLAMA_K_QUANTS=on
-//go:generate cmake --build ggml/build/cuda --target server --config Release
-//go:generate cmake -S gguf -B gguf/build/cuda -DLLAMA_CUBLAS=on -DLLAMA_ACCELERATE=on -DLLAMA_K_QUANTS=on
-//go:generate cmake --build gguf/build/cuda --target server --config Release
+//go:generate sh cuda.sh
```

Reply:
Let me drop out my go script, and you can add the bash script in a different commit. Seems like we're on the same page. Do note that `clinfo` needs a workaround because when OpenCL is not configured correctly, it hangs forever. There might be a more `pkg-config` like way to do this...

Reply:
Ah, there's an edge case. We probably want to have ROCm for GGUF and OpenCL for GGML in the ROCm case, so it gets a little funky. Testing locally, it doesn't hurt to enable both in the same llama binary, so that might be an option. I had to look up the difference between ROCM_HOME and ROCM_PATH, but it seems like they are both valid?

Reply:
Added a script.